infrastructure/src/main/java/org/corfudb/infrastructure/CompactorService.java

```diff
@@ -0,0 +1,162 @@
+     *
+     * @param interval interval to run the service
+     */
+    @Override
+    public void start(Duration interval) {
+        this.compactorLeaderServices = new CompactorLeaderServices(getCorfuRuntime(), serverContext.getLocalEndpoint());
+        this.corfuStore = new CorfuStore(getCorfuRuntime());
+        this.compactionTriggerPolicy.setCorfuRuntime(getCorfuRuntime());
+        if (getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis() > 0) {
+            this.compactionTriggerFreqMs = getCorfuRuntime().getParameters().getCheckpointTriggerFreqMillis();
+        }
+
+        orchestratorThread.scheduleWithFixedDelay(
+            this::runOrchestrator,
+            interval.toMillis(),
+            interval.toMillis(),
+            TimeUnit.MILLISECONDS
+        );
+    }
+
+    /**
+     * Invokes and cleans up the CorfuStoreCompactor jvm based on the status of CompactionManager
+     * Additionally, If the current node is the leader,
+     *      a. Invokes ValidateLiveness() to keep track of checkpointing progress by each client
+     *      b. Triggers the distributed compaction cycle based on the TriggerPolicy
+     */
+    private void runOrchestrator() {
+        boolean isLeader = isNodePrimarySequencer(updateLayoutAndGet());
+        compactorLeaderServices.setLeader(isLeader);
+        CheckpointingStatus managerStatus = null;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            managerStatus = (CheckpointingStatus) txn.getRecord(
+                    DistributedCompactor.COMPACTION_MANAGER_TABLE_NAME,
+                    DistributedCompactor.COMPACTION_MANAGER_KEY).getPayload();
+            txn.commit();
+        } catch (Exception e) {
+            syslog.warn("Exception in runOrchestrator: {}", e.getStackTrace());
+        }
+        syslog.trace("ManagerStatus: {}", managerStatus == null ? "null" : managerStatus.getStatus());
+        if (managerStatus != null) {
+            if (managerStatus.getStatus() == StatusType.FAILED || managerStatus.getStatus() == StatusType.COMPLETED) {
+                checkpointerJvmManager.shutdown();
+                checkpointerJvmManager.setIsInvoked(false);
+            } else if (managerStatus.getStatus() == StatusType.STARTED_ALL && !checkpointerJvmManager.isRunning()
+                    && !checkpointerJvmManager.isInvoked()) {
+                checkpointerJvmManager.invokeCheckpointing();
+            }
+        }
+
+        if (isLeader) {
+            if (managerStatus != null && (managerStatus.getStatus() == StatusType.STARTED ||
+                    managerStatus.getStatus() == StatusType.STARTED_ALL)) {
+                compactorLeaderServices.validateLiveness(livenessTimeout);
+            } else if (compactionTriggerPolicy.shouldTrigger(this.compactionTriggerFreqMs)) {
+                compactorLeaderServices.trimAndTriggerDistributedCheckpointing();
+                compactionTriggerPolicy.markCompactionCycleStart();
```

Comment:
shouldnt this be set before trimAndTriggerDistributedCheckpointing()?

Reply:
yes. good catch